#  Анализ базы данных сервиса он-лайн книг.

## Оглавление проекта

* [Краткое описание исследования](#Описание)
* [Подготовка данных](#Данные) 
* [Анализ базы данных](#Анализ)


## Краткое описание исследования <a id='Описание'></a>

Компания, приобрела сервис для чтения книг по подписке и нуждается в первом поверхостном анализе базы данных. Ниже несколько вопросов, на которые надо ответить в рамках исследования.

Описание данных и связи таблиц есть в [документе](https://disk.yandex.ru/i/OGG3gCoaUa-vog).

## Подготовка данных<a id='Данные'></a>

In [1]:
#импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
#устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя 
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], db_config['pwd'], db_config['host'], db_config['port'], db_config['db'])


In [3]:
#сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

### Посмотрим на таблицы в БД

In [4]:
q_books = ''' 
        SELECT *
        FROM books
        '''

q_authors = ''' 
        SELECT *
        FROM authors
        '''

q_publishers = ''' 
        SELECT *
        FROM publishers
        '''

q_ratings = ''' 
        SELECT *
        FROM ratings
        '''

q_reviews = ''' 
        SELECT *
        FROM reviews
        '''

In [5]:
#напишем функцию для отображения результата SQL запроса
def show_table(query):
    return pd.io.sql.read_sql(query, con = engine)

In [6]:
#отобразим первые строки таблиц и информацию по их содержимому
display(show_table(q_books).head())
display(show_table(q_books).info())
display(show_table(q_authors).head())
display(show_table(q_authors).info())
display(show_table(q_publishers).head())
display(show_table(q_publishers).info())
display(show_table(q_ratings).head())
display(show_table(q_ratings).info())
display(show_table(q_reviews).head())
display(show_table(q_reviews).info())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

### Вывод

В базе данных лежит 5 таблиц, связанных между собой по схеме из [полного описания задачи](https://disk.yandex.ru/i/OGG3gCoaUa-vog)

Таблицы не содержат пропусков и почти все имеют соответствующий формат данных, кроме даты в таблице _books_ что мы учтём при выполнении запроса. 

Приступаем к анализу.

## Анализ базы данных<a id='Анализ'></a>
Анализ будет состоять из пяти задач, описаных ниже. 

### Задача 1. Посчитайте, сколько книг вышло после 1 января 2000 года

**Подробнее:** посчитаем сумму id книг, дата публикации которых больше '01-01-2000' не включая эту дату. 

In [7]:
q_1 = ''' 
        SELECT COUNT(book_id)
        FROM books
        WHERE CAST(publication_date as date) > '01-01-2000'
        '''


display(show_table(q_1).head())

count
0    819

**Ответ:** После 1 января 2000 года вышло 819 книг. 

### Задача 2. Для каждой книги посчитайте количество обзоров и среднюю оценку
**Подробнее:** сджойним таблицы _books_, _ratings_ и _reviews_, посчитаем сумму `review_id` и средний `rating`. Отобразим данные по убыванию количества обзоров.

In [11]:
q_2 = ''' 
        SELECT b.book_id,
            b.title,
            COUNT(DISTINCT review_id) as rewiews_count,
            AVG(rating) as avg_rating
        FROM books AS b
        LEFT JOIN ratings AS rt ON rt.book_id = b.book_id 
        LEFT JOIN reviews AS rv ON rv.book_id = b.book_id
        GROUP BY b.book_id
        ORDER BY rewiews_count DESC;
        ''' 
  
display(show_table(q_2))

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     rewiews_count  avg_rating  
0                7    3.662500  
1                6    3.977273  
2                6    4.206897  
3                6    4.414634  
4                6    4.081081  
..             ...         ...  
995              0    3.666667  
996              0    3.000000  
997              0    5.000000  
998              0    4.000000  
999              0    4.000000  

[1000 rows x 4 columns]

**Ответ:** Мы составили таблицу из 1000 строк(соответствует количеству книг), в которой посчитали количество обзоров и средние оценки, отсортировав по количеству обзоров. Судя по всему, высокое количество обзоров не гарантирует высокой оценки: на первом месте у нас получилась книга с рейтингом 3.66.

### Задача 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц (так вы исключите из анализа брошюры)
**Подробнее:** сджойним таблицы _books_ и _publishers_, выберем те строки, где количество страниц больше 50 и сгруппируем по имени издательства. Отсортируем по количеству книг и отобразим только первую строку.

In [ ]:
q_3 = ''' 
        SELECT publisher,
               COUNT(b.publisher_id) as book_count
               
        FROM books AS b
        RIGHT JOIN publishers AS pb ON pb.publisher_id = b.publisher_id
        WHERE b.num_pages > 50
        GROUP BY publisher
        ORDER BY book_count DESC
        LIMIT 1

        ''' 
       
display(show_table(q_3))

**Ответ:** Больше всего книг с 50+ страниц(42) выпустило издетельство 'Penguin Books'.

### Задача 4. Определите автора с самой высокой средней оценкой книг (учитывайте только книги с 50 и более оценками)
**Подробнее:** сджойним таблицы _books_, _ratings_ и _authors_, сгруппируем по `author_id`, оставим только те строки, в которых сумма `rating_id` больше 50, отсортируем по убыванию по `mean_rating` и выведем только первую строку.

In [ ]:
q_4 = ''' 
        SELECT author,
               AVG(rating) as mean_rating
        FROM books AS b
        RIGHT JOIN ratings AS rt ON rt.book_id = b.book_id
        RIGHT JOIN authors AS a ON a.author_id = b.author_id
        GROUP BY a.author_id

        HAVING COUNT(rating_id) > 50 
        ORDER BY mean_rating DESC
        LIMIT 1

        ''' 
  
display(show_table(q_4))

**Ответ:** автор с самой высокой средней оценкой книг(4.29) - не автор, а авторы: J.K. Rowling/Mary GrandPré	

### Задача 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.
**Подробнее:** сделаем первый подзапрос: достанем из таблицы _ratings_ `username` пользователей, для которых после группировки по юзернейму сумма `rating_id` будет больше 50. Сделаем второй подзапрос: используя пользователей из первого подзапроса(их 6), выберем из них только те, которые пересекутся со сгруппированой по `username` таблицей _reviews_ и отобразим их среднее количество обзоров. После чего, сделаем основной запрос, в котором посчитаем среднее количество отзывов полученых двумя подзапросами и выведем округлив до целого. 

In [ ]:
#q_5_1 = ''' 
        #SELECT username
        #FROM ratings
        #GROUP BY username
        #HAVING COUNT(rating_id) > 50
#        ''' 


#q_5_2 = ''' 
        #SELECT username,
               #COUNT(review_id) AS mean_review
        #FROM reviews
        #WHERE username IN (SELECT username
                          #FROM ratings
                          #GROUP BY username
                          #HAVING COUNT(rating_id) > 50)
        #GROUP BY username
#       ''' 

q_5_3 = ''' 
        SELECT ROUND(AVG(mean_review)) AS total_mean_review
        FROM (SELECT username,
                     COUNT(review_id) as mean_review
              FROM reviews
              WHERE username IN (SELECT username
                                 FROM ratings
                                 GROUP BY username
                                 HAVING COUNT(rating_id) > 50)
              GROUP BY username) AS review_info
        ''' 


display(show_table(q_5_3))


**Ответ:** пользователи, которые поставили больше 50 оценок, в среднем пишут 24 обзора. 

### Вывод:
По итогам нашего обзора можем заключить, что база данных хранит информацию о 1000 книг, их авторах(636). Книги опубликованы 340-ми издательствами, у нас есть так же данные о 2793-ех обзорах и 6456-ти оценках книг.

Данные не имеют пропусков и в общем хорошего качества. 

Мы сделали несколько расчетов, определили, что:
- в базе 819 книг и брошюр выпущенных после 2000 года;
- книга Twilight (Twilight #1) имеет самое большое число обзоров - 7 и средний рейтинг 3.66;
- больше всего книг с 50+ страниц(42) выпустило издетельство 'Penguin Books';
- авторы с самой высокой средней оценкой книг(4.29): J.K. Rowling/Mary GrandPré;
- пользователи, которые поставили больше 50 оценок, в среднем пишут 24 обзора.

Данные выглядят поными и адекватными для дальнейших анализов.